<html lang="eng">
    <head>
        <!-- Is the title of the page, over the web page-->
        <title>BradBand</title> 
    </head>
    <body>
        
        <h1> BroadBand</h1>
        <p>  BroadBand is a web-based resource developed with the Shiny R framework, designed for the exploration of genomic data selected among a comprehensive selection of breast cancer (BRCA) studies with available information on somatic single nucleotide variants (SNVs) and somatic copy number alterations (SCNAs). BroadBand allows an intuitive and custom navigation of SNVs and SCNAs data which can be filtered and summarized at different resolution levels, from gene to chromosome, and stratified across the different breast cancer histotypes and tumor classifications.</p>
        <!-- <div align="center">Hello World!</div> -->

        <h2>Cohort overview and samples stratification</h2>
        <p>The “Overview” panel, the home page of the BroadBand resource, provides a detailed report (as bar
            chart and table) about the number of samples, carefully stratified by molecular features and data type
            availability, included in the selected cohort(s).</p>
        <p>The widget panel (Figure 1A) is populated with multiple selection options which will update the
            BRCA data displayed by default. Specifically, the widget panel allows the user to:
            1. Upload a new BRCA dataset to be inspected in addition to those ones available by default.
            The cancel button will reverse the operation by removing the uploaded dataset among
            available data.
            2. Select samples based on somatic aberrations data of interest. For example, the user can select
            only those samples with SCNAs and/or somatic (non-synonymous) SNVs data available.
            153. Select BRCA cohorts of interest. If the user uploaded a custom dataset, it will be listed here.
            4. Select data based on BRCA subtypes.
            5. Select data based on tumor classification.
            6. Reset all settings to default.
            7. Apply selected options to data and update plots and the table accordingly.
            In addition to the widgets bar, the page layout also includes two plots, a table and three downloads
            buttons which allow the user to save all outputs displayed. The plots and the table are the results of
            options selected by the user. Specifically, the first bar chart reports the number of samples stratified
            by cancer subtypes and tumor classification, the second plot is composed by two bar charts, one for
            each tumor class (primary and metastatic), and reports sample counts by cancer subtypes. The table
            (Figure 1C) reports data represented in all plots</p>
        <h2>Genes aberrations frequencies based on somatic SNVs data</h2>
        <p>The second panel is focused on somatic single nucleotide variants data and provides a comprehensive
            report about the frequencies of aberration of genes </p>
        <p>The widget panel (Figure 2A) is populated with multiple selection options which will update the
            BRCA data displayed by default. Specifically, the widget panel allows the user to:
            1. Upload a new BRCA dataset to be inspected in addition to those ones available by default.
            The cancel button will reverse the operation by removing the uploaded dataset among
            available data.
            2. Select BRCA cohorts of interest. If the user uploaded a custom dataset, it will be listed here.
            3. Select data based on BRCA subtypes.
            4. Select data based on tumor classification.
            5. Select the N most frequently aberrant genes by using the provided slider.
            6. Reset all settings to default.
            7. Apply selected options to data and update plots and the table accordingly.
            In addition to the widgets bar, the page layout also includes two plots, a table and three downloads
            buttons which allow the user to save all outputs displayed. The plots and the table are the results of
            options selected by the user. Specifically, the first plot (Figure 2B, left plot) comprises multiple bar
            charts, data represented are stratified by tumor classification (one for each row) and for breast cancer
            subtypes (one for each column; the first column “all” comprises tumor subtypes data altogether) as
            reported on top of each inset panel. Within each bar chart, genes (listed on the y-axis) are ranked
            based on their aberration frequency.
            Alongside (Figure 2B, right plot), the heatmap compares aberration frequencies of genes between
            primary and metastatic and by cancer subtypes. Data represented in this way could be helpful to spot
            by visual inspection which genes are mutually exclusive within a single tumor class. The table
            (Figure 2C) reports data represented in both plot</p>
        <h2>Chromosomal aberrations frequencies based on SCNAs data</h2>
        <p>The third panel is focused on somatic copy number aberrations data, this panel provides a full report
            about the frequencies of somatic copy number aberrations occurring in genomic regions of different
            ranges. Specifically, aberrations occurring in gene, chromosomal cytobands and chromosomal arms,
            are computed based on molecular features, tumor classification of available data included in the
            selected cohorts. </p>
        <p>The widget panel (Figure 3A) is populated with multiple selection options which will update the BRCA data displayed by default. Specifically, the widget panel allows the user to:
            1. Upload a new BRCA dataset to be inspected in addition to those ones available by default.
            The cancel button will reverse the operation by removing the uploaded dataset among
            available data.
            2. Select BRCA cohorts of interest. If the user uploaded a custom dataset, it will be listed here.
            3. Select data based on BRCA subtypes.
            4. Select data based on tumor classification.
            5. Set granularity of SCNAs event. The user can select to compute and report the frequencies of
            aberrations based on 2 classes of aberrations: amplifications and deletion events or,
            alternatively, on 4 classes:
            homozygous deletion, hemizygous deletion, gain, and
            amplification events.
            6. Select the SCNAs event of interest to be displayed.
            7. Apply a threshold on aberration frequencies by using the provided slider.
            8. Select the chromosome of interest to be displayed.
            9. Select the cytoband of interest to be displayed.
            10. Reset all settings to default.
            11. Apply selected options to data and update plots and tables accordingly.
            In addition to the widgets bar, the page layout also includes two plots, a table and five downloads
            buttons which allow the user to save all outputs displayed. The plots and the tables are the results of
            all options selected by the user. The first plot is a set of bar charts, one for each chromosome, to
            provide a landscape of SCNAs frequencies of aberrations. In the second plot the user can select a
            chromosome of interest to get a closer look at the frequencies of aberration across the chromosome
            cytobands. The third plot reports, within a selected cytoband of interest, the frequency of aberrations
            at the gene level. A gene is highlighted (Figure 3D) if it is a known cancer driver gene. Genomic
            location information (Figure 3B, 3C) and BRCA dataset used to compute frequencies of aberrations
            displayed (Figure 3C, 3D) are reported in the plot legend. </p>
            <div id="Page1" class="page" style="">
                Content of page 1
              </div>
              <div id="Page2" class="page" style="display:none">
                Content of page 2
              </div>
              <div id="Page3" class="page" style="display:none">
                Content of page 3
              </div>
    </body>
</html>
